src/lib/yaraparse.py

```diff
@@ -73,174 +145,206 @@ def ursify_string(string) -> Optional[str]:
-    return [s for s in ursa_strings if s is not None]
-
-
-def expand_set_expression(children: SetExpression, rule_strings) -> List[str]:
-    parsed_elements = []
-    for e in children.elements:
-        if type(e) is StringWildcardExpression:
-            parsed_elements += expand_string_wildcard(e, rule_strings)
-        elif type(e) is StringExpression:
-            parsed_elements.append(str_expr(e, rule_strings))
+        if left and right:
+            return UrsaExpression.or_(left, right)
+        elif not left and not right:
+            return None
         else:
-            print(f"Unknown set expression type: {type(e)}")
-            parsed_elements.append("")
-
-    return parsed_elements
+            return left or right
+
+    def pare_expr(
+        self, condition: ParenthesesExpression
+    ) -> Optional[UrsaExpression]:
+        return self.traverse(condition.enclosed_expr)
+
+    def str_expr(
+        self, condition: StringExpression
+    ) -> Optional[UrsaExpression]:
+        return ursify_string(self.strings[condition.id])
+
+    def expand_string_wildcard(
+        self, condition: StringWildcardExpression
+    ) -> List[UrsaExpression]:
+        condition_regex = re.escape(condition.text)
+        condition_regex = condition_regex.replace("\\*", ".*")
+        filtered_strings = [
+            v for k, v in self.strings.items() if re.match(condition_regex, k)
+        ]
+
+        ursa_strings = [ursify_string(x) for x in filtered_strings]
+        return [s for s in ursa_strings if s is not None]
+
+    def expand_set_expression(
+        self, children: SetExpression
+    ) -> List[Optional[UrsaExpression]]:
+        parsed_elements: List[Optional[UrsaExpression]] = []
+        for e in children.elements:
+            if type(e) is StringWildcardExpression:
+                parsed_elements += self.expand_string_wildcard(e)
+            elif type(e) is StringExpression:
+                parsed_elements.append(self.str_expr(e))
+            else:
+                raise RuntimeError(f"Unknown set expression: {type(e)}")
+
+        return parsed_elements
+
+    def of_expr(self, condition: OfExpression) -> Optional[UrsaExpression]:
+        how_many = condition.text[: condition.text.find("of")].strip()
+        counter = None
+
+        children = condition.iterated_set
+
+        if type(children) is SetExpression:
+            all_elements = self.expand_set_expression(children)
+        elif type(children) is ThemExpression:
+            all_elements = [ursify_string(k) for k in self.strings.values()]
+        else:
+            raise YaraParseError(f"Unsupported of_expr type: {type(children)}")
 
+        parsed_elements = [e for e in all_elements if e is not None]
 
-def of_expr(condition, rule_strings) -> Optional[str]:
-    how_many = condition.text[: condition.text.find("of")].strip()
-    counter = None
+        if how_many == "all":
+            counter = len(all_elements)
```

Comment:
I didn't notice this in the last review but shouldn't this be `len(parsed_elements)`?

Otherwise if we have `all of (x, y, z)` and `y` turns out to be a None in the ursa query, won't the final query look like `min 3 of (x, y)`?

Reply:
:thinking: that's tricky.

First of all, the change was intentional. But now I think it was not correct after all.

Let's consider three cases and what they should be simplified to:

```
all of ("aaa", "bbb", /regex/)  =>  (equivalent to) 2 of ("aaa", "bbb")?
```

```
3 of ("aaa", "bbb", /regex/)  =>  (equivalent to) 2 of ("aaa", "bbb")?
```

```
2 of ("aaa", "bbb", /regex/)  =>  (equivalent to) 1 of ("aaa", "bbb")?
```

In other words, I think using `all` as an alias for `len(all_elements)` here is correct, but I should also subtract number_of_none_elements from counter *in all cases*.

(ofc remember to cap counter at 0 from the bottom)

Please read my thoughts and tell me what you think :thinking:.

Reply:
So for `how_many == "all"` `len(all_elements) - number_of_none_elements` should be equal to `len(parsed_elements)` right?

But yes, subtracting it for all cases makes sense